File: leecode/code/10/10.java
class Solution {

    public boolean isMatch(String s, String p) {
        if (p.isEmpty()) return s.isEmpty();

        boolean flag = !s.isEmpty() &&
                        (s.charAt(0) == p.charAt(0) || p.charAt(0) == '.');

        if (p.length() >= 2 && p.charAt(1) == '*') 
            return (isMatch(s, p.substring(2)) || (flag && isMatch(s.substring(1), p)));
        else 
            return (flag && isMatch(s.substring(1), p.substring(1)));
    }
}

File: leecode/code/10/10.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        if len(p) == 0:
            return len(s) == 0
        else: 
            flag = len(s) > 0 and (s[0] == p[0] or p[0] == '.')


            if len(p) > 1 and p[1] == '*':
                return self.isMatch(s, p[2:]) or (flag and self.isMatch(s[1:], p))
            
            else:
                return (flag and self.isMatch(s[1:], p[1:]))